File: leetcode/array/283. Move Zeroes/py/main.py
class Solution:
    # Time O(N)
    # Space O(1)
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = -1  # index 0
        n = len(nums)

        for idx, v in enumerate(nums):
            if v == 0:
                i = idx
                break

        if i + 1 >= n or i == -1:
            return

        j = i + 1
        while j < n:
            if nums[i] != nums[j]:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
                j += 1
            else:
                j += 1

File: leetcode/array/283. Move Zeroes/js/main.js
// Time O(N)
// Space O(1)
const moveZeroes = nums => {
  let n = nums.length;
  let cnt = 0;

  for (let i = 0; i < n; i++) {
    if (nums[i] == 0) {
      cnt++;
    }

    let j = i;
    while (cnt > 0 && j < n) {
      if (nums[j] != 0) {
        swap(nums, i, j);
        cnt--;
      }
      j++;
    }

    if (j == n) {
      break;
    }
  }
};

function swap(nums, i, j) {
  return ([nums[i], nums[j]] = [nums[j], nums[i]]);
}

// Time O(N)
// Space O(1)
const moveZeroes_II = nums => {
  let n = nums.length;
  let nxt = 0;

  for (let num of nums) {
    if (num != 0) {
      nums[nxt] = num;
      nxt++;
    }
  }

  for (let i = nxt; i < n; i++) {
    nums[i] = 0;
  }
};
